File: Cpp/238_Product_of_Array_Except_Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> result(nums.size(), 1);
        int pre = nums.front();
        for(int i=1; i<nums.size(); ++i){
            result[i] *= pre;
            pre *= nums[i];
        }
        
        int post = nums.back();
        for(int i=nums.size()-2; i>=0; --i){
            result[i] *= post;
            post *= nums[i];
        }
        
        return result;
    }
};

File: Kotlin/238_Product_of_Array_Except_Self.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        val n = nums.size
        val result = IntArray(nums.size) { 1 }
        
        var pre = 1
        var post = 1
        for (i in nums.indices) {
            result[i] *= pre
            pre *= nums[i]
            
            result[n - i - 1] *= post
            post *= nums[n - i - 1]
        }
        
        return result
    }
}